### DISCUSSION TITLE
Test Framework like Spring Boot Test

### DISCUSSION THREAD

**Discussion by Orchius83** [serverpod/serverpod#1007] [Ideas]
Morning everyone, I'm using Serverpod a lot and now I'm starting a big project on it. 
I'm feeling that I miss the latest step to be totally productive that is the ability to test with a Mocked Session to save and retrieve Entities in my tests. 
Is there anything that is in development right now?
Thank

**Comment by Isakdl** [comment]
I fully agree, had the same thought recently. I think it is a rather important feature and we will add it in the future!

**Reply by Isakdl** [reply]
I created an issue to track this, [here](https://github.com/serverpod/serverpod/issues/1008).

**Reply by Orchius83** [reply]
If you want I can do a first implementation with an in memory database? or you were thinking of something else?

**Reply by Isakdl** [reply]
I haven't thought too much about how to deal with the database, however, it is a none trivial thing to create an in-memory database. There are a lot of features to cover and it is important to be as close to the real thing as possible. I think it might be easier to solve by simply letting the developer mock the needed `insert`, `update`, `find`, etc methods on a per endpoint basis rather than building an entire fake database. Or run it against a real DB in a container (this adds some complexity and potential interdependency in the tests). Even allowing both setups could potentially be cool.

But if you want to take a stab at it sure! Perhaps it would be easier to start with handling the auth part instead of the database.

The most important part here though is that there is a clean and easy-to-use interface.

**Reply by Orchius83** [reply]
absolutely, I want it to be totally transparent for end user, I will have a go and be back with a PoC

**Reply by Frank1016** [reply]
Isn't the Staging database for this kind of testing purpose?
